### PR TITLE
Add option to read hardware timestamps.

### PR DESCRIPTION
**Pull request by FabianInostroza**
I took the code from https://marc.info/?l=linux-can&m=138244457221323&w=2.

Unfortunately only the pcan driver provides hardware timestamps.

Note that I don't own a PCAN adapter, I tested this modification using a self made pcan compatible adapter, receiving one message every 400 us and the hardware timestamp sometimes jumps back or forward. The same self made adapter gives correct timestamps in Windows with the official driver and PCAN-view.

### CODE REVIEW

File: candump.c [hunk 1]
```diff
@@ -688,8 +707,16 @@ int main(int argc, char **argv)
...
 						memcpy(&tv, CMSG_DATA(cmsg), sizeof(tv));
+					} else if (cmsg->cmsg_type == SO_TIMESTAMPING) {
+
+						struct timespec *stamp = (struct timespec *)CMSG_DATA(cmsg);
+
+						stamp += 2;
+						tv.tv_sec = stamp->tv_sec;
+						tv.tv_usec = stamp->tv_nsec/1000;
+					}
 					else if (cmsg->cmsg_type == SO_RXQ_OVFL)
```
Comment (marckleinebudde) [inline candump.c]: Please make it:
`} else if () {`

File: candump.c [hunk 2]
```diff
@@ -688,8 +707,16 @@ int main(int argc, char **argv)
...
 				     cmsg && (cmsg->cmsg_level == SOL_SOCKET);
 				     cmsg = CMSG_NXTHDR(&msg,cmsg)) {
-					if (cmsg->cmsg_type == SO_TIMESTAMP)
+					if (cmsg->cmsg_type == SO_TIMESTAMP) {
 						memcpy(&tv, CMSG_DATA(cmsg), sizeof(tv));
+					} else if (cmsg->cmsg_type == SO_TIMESTAMPING) {
+
+						struct timespec *stamp = (struct timespec *)CMSG_DATA(cmsg);
+
+						stamp += 2;
```
Comment (hartkopp) [inline candump.c]: This offset calculation might be clear when looking at the referenced code or in the Linux code.
But "stamp += 2;" should be commented and properly referenced WHY we can find the hardware timestamp at this point.

Btw. thanks for the patch :-)

**Comment by FabianInostroza** [comment]
Hi, it's done.

I left candump logging to a file for a long time (15 min aprox) and the timestamp wraps at a certain point (not constant). I don't think this is related to the adapter I'm using.

**Comment by FabianInostroza** [comment]
Hi,

The problem with the timestamp jumps was in the firmware, the cause of the wrap after a long time was oscillator frequency error (15 secs every 4200 sec).

Is it possible for the kernel to reorder the messages before being sent to the clients?
Now I have noticed that candump show some messages reordered, look at the following candump log 
where the first four data bytes contain a sequence number
```
(1512440012.199080) can0 1F0#7269000005060708
(1512440012.199946) can0 1F1#7369000005060708
(1512440012.199946) can0 00ABCDEF#7469000005060708 
(1512440012.201872) can0 1F0#7869000005060708 ---> sequence number jumps
(1512440012.201872) can0 1F1#7969000005060708
(1512440012.201872) can0 00ABCDEF#7A69000005060708
(1512440012.201013) can0 1F0#7569000005060708 ---> sequence number jumps
(1512440012.201014) can0 1F1#7669000005060708
```
Also look at the timestamps, they also are out of order.
I verified the usb packets that the device sends (using usbmon and wireshark) and the messages 
are sent in correct order and timestamps are correct.
Below is a capture taken at the same time using hardware timestamps
```
(0000005020.582582) can0 1F0#7269000005060708
(0000005020.582923) can0 1F1#7369000005060708
(0000005020.583307) can0 00ABCDEF#7469000005060708
(0000005020.584672) can0 1F0#7869000005060708
(0000005020.585014) can0 1F1#7969000005060708
(0000005020.585440) can0 00ABCDEF#7A69000005060708
(0000005020.583648) can0 1F0#7569000005060708
(0000005020.583990) can0 1F1#7669000005060708
```
I used the script can_logger.py provided by python-can and also show this behavior.

**Comment by hartkopp** [comment]
Can you tell more about the used CAN hardware? Is it USB based?
See https://marc.info/?l=linux-can&m=148007442317274&w=2
Can we discuss this - not candump-related - issue at linux-can mailing list?
Tnx.

**Comment by FabianInostroza** [comment]
Yes, it is USB based. I didn't found that thread when googling.
I will try the suggestions posted in that thread.
Do you want me to include some additional information?

**Comment by hartkopp** [comment]
I personally used a PCAN USB pro FD to test your support for hardware timestamps.
You might just subscribe to the linux-can mailing list, see:
http://vger.kernel.org/vger-lists.html#linux-can 

The mailing list does not have much traffic.

**Comment by FabianInostroza** [comment]
> I personally used a PCAN USB pro FD to test your support for hardware timestamps.
> You might just subscribe to the linux-can mailing list, see:
> http://vger.kernel.org/vger-lists.html#linux-can
> 
> The mailing list does not have much traffic.

@hartkopp Hi, I subscribed to the mailing list and sent a email with a patch but it isn't shown in the archives, I've received mails sent by other members but no error or notification about my own email.
What could be wrong?

**Comment by hartkopp** [comment]
Did you send it with `git send-email`?
Usually only HTML mails are inappropriate and cause problems.

**Comment by FabianInostroza** [comment]
> Did you send it with `git send-email`?
> Usually only HTML mails are inappropriate and cause problems.

Mmm, that could be the cause. I sent the email using gmail web client and after the mail was sent I noticed that the format of the patch was messed up and the mail was sent with some html.
Will try again using plain text.

**Pull request merged**